// fs/index.js
// node_modules/fs/index.js
const bridge =
    (typeof Polyglot !== 'undefined' && typeof Polyglot.import === 'function')
        ? Polyglot.import('fsBridge')
        : (typeof globalThis !== 'undefined' ? globalThis.fsBridge : undefined);

if (!bridge) throw new Error("fsBridge not available (enable polyglot or inject globalThis.fsBridge)");

// ---- Sync API ----
function readFileSync(path, options) {
    const enc = normalizeEncoding(options);
    if (enc && enc !== 'utf8') throw new Error("fs shim supports only 'utf8' encoding");
    return bridge.readFileUtf8(String(path));
}

function writeFileSync(path, data, options) {
    const enc = normalizeEncoding(options);
    if (enc && enc !== 'utf8') throw new Error("fs shim supports only 'utf8' encoding");
    if (typeof data !== 'string') data = String(data);
    bridge.writeFileUtf8(String(path), data);
}

function existsSync(path) {
    try { return !!bridge.exists(String(path)); }
    catch { return false; }
}

function readdirSync(path) {
    return Array.from(bridge.readdir(String(path)));
}

function statSync(path) {
    const s = bridge.stat(String(path));
    return {
        size: s.size,
        mtimeMs: s.mtimeMs,
        isFile: () => !!s.isFile,
        isDirectory: () => !!s.isDir,
    };
}

// ---- Minimal async wrappers (Promise-based) ----
function readFile(path, options, cb) {
    const enc = normalizeEncoding(options);
    const run = () => readFileSync(path, enc || 'utf8');
    if (typeof cb === 'function') { try { cb(null, run()); } catch (e) { cb(e); } return; }
    return new Promise((res, rej) => { try { res(run()); } catch (e) { rej(e); } });
}

function writeFile(path, data, options, cb) {
    const enc = normalizeEncoding(options);
    const run = () => writeFileSync(path, data, enc || 'utf8');
    if (typeof cb === 'function') { try { cb(null, run()); } catch (e) { cb(e); } return; }
    return new Promise((res, rej) => { try { res(run()); } catch (e) { rej(e); } });
}

function normalizeEncoding(opt) {
    if (!opt) return 'utf8';
    if (typeof opt === 'string') return opt.toLowerCase();
    if (typeof opt === 'object' && opt.encoding) return String(opt.encoding).toLowerCase();
    return 'utf8';
}

module.exports = {
    readFileSync, writeFileSync, existsSync, readdirSync, statSync,
    readFile, writeFile
};
