// fetch/index.js
// node_modules/fetch/index.js
const bridge =
    (typeof Polyglot !== 'undefined' && typeof Polyglot.import === 'function')
        ? Polyglot.import('httpBridge')       // from ctx.getPolyglotBindings.putMember(...)
        : (typeof globalThis !== 'undefined' ? globalThis.httpBridge : undefined);

if (!bridge) throw new Error("httpBridge not available");

function fetchShim(url, opts = {}) {
    const method = (opts.method || 'GET').toUpperCase();
    const headers = opts.headers || {};
    const body =
        typeof opts.body === 'string' ? opts.body :
            (opts.body == null ? '' : JSON.stringify(opts.body));
    const timeoutMs = typeof opts.timeoutMs === 'number' ? opts.timeoutMs : 10000;

    return new Promise((resolve, reject) => {
        try {
            const r = bridge.fetchText(url, method, headers, body, timeoutMs); // @HostAccess.Export
            const res = {
                ok: r.getStatus() >= 200 && r.getStatus() < 300,
                status: r.getStatus(),
                headers: r.getHeaders ? r.getHeaders() : {},
                text: () => Promise.resolve(r.getBody()),
                json: () => Promise.resolve(JSON.parse(r.getBody())),
            };
            resolve(res);
        } catch (e) { reject(e); }
    });
}

module.exports = fetchShim;

// also expose as a global so `require('fetch'); fetch(...)` works
if (typeof globalThis !== 'undefined' && typeof globalThis.fetch === 'undefined') {
    globalThis.fetch = fetchShim;
}
